graphql_compiler/compiler/emit_sql.py

```diff
@@ -143,35 +165,71 @@ class SQLFoldObject(object):
+            primary_key: primary_key of the vertex immediately outside the fold. Used to set the
+                        group by as well as join the fold subquery to the rest of the query.
+                        Composite keys unsupported.
         """
         # table containing output columns
         # initially None because output table is unknown until call to visit_output_vertex
         self._output_vertex_alias = None
 
         # table for vertex immediately outside fold
         self._outer_vertex_alias = outer_vertex_table
+        if len(primary_key.columns) > 1:
+            raise NotImplementedError(u'Composite keys not supported. '
+                                      u'A composite primary key {} was found for table {}. '
+                                      u'SQL fold only supports non-composite primary '
+                                      u'keys'.format(primary_key, outer_vertex_table.original))
+
+        # name of the field used in the primary key for the vertex outside the fold
+        # current implementation does not support composite primary keys
+        #
+        # we must use the name of the column as opposed to the column itself because
+        # primary key refers to the column from the original table, while we need the
+        # identically named column from its alias
+        self._outer_vertex_primary_key = primary_key.columns_autoinc_first[0].description
 
         # group by column for fold subquery
-        self._group_by = [self._outer_vertex_alias.c[join_descriptor.from_column]]
+        self._group_by = [self._outer_vertex_alias.c[self._outer_vertex_primary_key]]
 
-        # List of 3-tuples describing the join required for each traversal in the fold
+        # List of SQLFoldTraversalDescriptor namedtuples describing each traversal in the fold
         # starting with the join from the vertex immediately outside the fold to the folded vertex:
-        #
-        #  edge: join descriptor for the columns used to join one vertex to the next in the fold
-        #  from_table: the table on the left side of the join
-        #  to_table: the table on the right side of the join
-        self._join_info = []
-
+        self._traversal_descriptors = []
         self._outputs = []  # output columns for fold
 
+        self._ended = False  # indicates whether `end_fold` has been called on this object
+
+    def __str__(self):
+        """Produce string used to customize error messages."""
+        if self._outer_vertex_alias is None:
+            return u'Invalid fold: no vertex preceding fold.'
+        elif self._output_vertex_alias is None:
+            return u'Vertex outside fold: {}. Output vertex for fold: None'.format(
+                self._outer_vertex_alias.original
+            )
+        else:
+            return u'Vertex outside fold: {}. Output vertex for fold: {}'.format(
+                self._outer_vertex_alias.original, self._output_vertex_alias.original
+            )
```

Comment:
I don't know if these messages will read clearly in the error messages that use these methods. Ex: these will be put in the middle of sentences, but are capitalized and won't really complete the sentences.